File: com/dev/leetcodehard/BasicCalculatorIII.java
/*
https://leetcode.com/problems/basic-calculator-iii/

status - completed
 */

package com.dev.leetcodehard;

import com.dev.logger.basePrinter;

class BasicCalculatorIII extends basePrinter{

    static int pos;
    static String expr;

    static int calculate(String s){
        expr = s.replaceAll("\\s+", "");
        pos = 0;
        return parseExpression();
    }

    static int parseExpression(){
        int result = parseTerm();
        while (pos < expr.length() && (expr.charAt(pos) == '+' || expr.charAt(pos) == '-')){
            char op = expr.charAt(pos);
            pos++;
            int term = parseTerm();
            if (op == '+'){
                result += term;
            }
            else{
                result -= term;
            }
        }
        return result;
    }

    static int parseTerm(){
        int result = parseFactor();
        while (pos < expr.length() && (expr.charAt(pos) == '*' || expr.charAt(pos) == '/')){
            char op = expr.charAt(pos);
            pos++;
            int factor = parseFactor();
            if (op == '*'){
                result *= factor;
            }
            else{
                result /= factor;
            }
        }
        return result;
    }

    static int parseFactor(){
        if (expr.charAt(pos) == '-'){
            pos++;
            return -parseFactor();
        }
        if (expr.charAt(pos) == '+'){
            pos++;
            return parseFactor();
        }
        if (expr.charAt(pos) == '('){
            pos++;
            int result = parseExpression();
            pos++;
            return result;
        }

        int start = pos;
        while (pos < expr.length() && Character.isDigit(expr.charAt(pos))){
            pos++;
        }
        return Integer.parseInt(expr.substring(start, pos));
    }

    public static void main(String[] args) {
        logp("input 1+1 expected 2 actual " + calculate("1+1"));
        logp("input 6-4/2 expected 4 actual " + calculate("6-4/2"));
        logp("input 2*(5+5*2)/3+(6/2+8) expected 21 actual " + calculate("2*(5+5*2)/3+(6/2+8)"));
        logp("input (2+6*3+5-(3*14/7+2)*5)+3 expected -12 actual " + calculate("(2+6*3+5-(3*14/7+2)*5)+3"));
    }
}

File: typescript/leetcodehard/basicCalculatorIII.ts
/*
https://leetcode.com/problems/basic-calculator-iii/

status - completed
*/

import { logp } from "../utils/logger";

function calculate(s: string): number {
  let pos = 0;

  function peek(): string {
    return s[pos];
  }

  function parseExpression(): number {
    let result = parseTerm();

    while (pos < s.length && (peek() === "+" || peek() === "-")) {
      const op = peek();
      pos++;
      const term = parseTerm();
      result = op === "+" ? result + term : result - term;
    }

    return result;
  }

  function parseTerm(): number {
    let result = parseFactor();

    while (pos < s.length && (peek() === "*" || peek() === "/")) {
      const op = peek();
      pos++;
      const factor = parseFactor();
      result = op === "*" ? result * factor : Math.trunc(result / factor);
    }

    return result;
  }

  function parseFactor(): number {
    while (pos < s.length && peek() === " ") pos++;

    let sign = 1;
    if (peek() === "+" || peek() === "-") {
      sign = peek() === "-" ? -1 : 1;
      pos++;
      while (pos < s.length && peek() === " ") pos++;
    }

    let value: number;
    if (peek() === "(") {
      pos++;
      value = parseExpression();
      while (pos < s.length && peek() === " ") pos++;
      pos++;
    } else {
      let numStr = "";
      while (pos < s.length && /[0-9]/.test(peek())) {
        numStr += peek();
        pos++;
      }
      value = Number(numStr);
    }

    while (pos < s.length && peek() === " ") pos++;
    return sign * value;
  }

  return parseExpression();
}

logp(`input "1+1" expected 2 actual ${calculate("1+1")}`);
logp(`input "6-4/2" expected 4 actual ${calculate("6-4/2")}`);
logp(`input "2*(5+5*2)/3+(6/2+8)" expected 21 actual ${calculate("2*(5+5*2)/3+(6/2+8)")}`);
